### PR TITLE
Trading competitions support

### PR DESCRIPTION
**Pull request by batrudinych**

### CODE REVIEW

**Comment by robertkowalski** [comment]
tests are failing...

File: lib/routes-pub.js [hunk 1]
```diff
@@ -7,19 +7,28 @@ const { TextDecoder, TextEncoder } = require('util')
...
+  sendGenericError
 } = require('./utils')
 
 class RoutesPub {
   constructor (conf, plugins) {
     this.conf = conf
     this.plugins = plugins
 
+    this.onTimeHttpRequest = this.onTimeHttpRequest.bind(this)
+    this.onTosHttpRequest = this.onTosHttpRequest.bind(this)
```
Comment (robertkowalski) [inline lib/routes-pub.js]: why rebinding Tos handlers etc here?

File: lib/moonbeam.js [hunk 2]
```diff
@@ -103,21 +103,24 @@ class Moonbeam {
 
     app.options('*', cmw)
 
+    app.get('/competition/leaderboard/vol', cmw, privRoutes.onCompetitionVolLbHttpRequest.bind(privRoutes))
+    app.post('/competition/signup', cmw, privRoutes.onCompetitionSignUpHttpRequest.bind(privRoutes))
```
Comment (robertkowalski) [inline lib/moonbeam.js]: we shoudl either go for binding in the handlers, like you did in the other code at L115, or do the bind call here.

File: lib/gr-client.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+const Peer = Grenache.PeerRPCClient
+
+const link = new Link(conf.link)
+link.start()
+
+const peer = new Peer(link, {})
+peer.init()
+
+exports.notifySignup = ({ user }, cb) => {
```
Comment (robertkowalski) [inline lib/gr-client.js]: you can use https://github.com/bitfinexcom/bfx-lib-server-js

File: config/moonbeam.mongo.conf.json.example [hunk 4]
```diff
@@ -2,5 +2,7 @@
   "dbName": "foo_history",
   "mongoUrl": "mongodb://localhost",
   "collection_trades": "userdata",
-  "collection_tos": "tos"
+  "collection_tos": "tos",
+  "collection_competitions": "competitions",
```
Comment (robertkowalski) [inline config/moonbeam.mongo.conf.json.example]: we can remove it, right?

File: lib/moonbeam.js [hunk 5]
```diff
@@ -103,21 +103,24 @@ class Moonbeam {
 
     app.options('*', cmw)
 
+    app.get('/competition/leaderboard/vol', cmw, privRoutes.onCompetitionVolLbHttpRequest.bind(privRoutes))
```
Comment (robertkowalski) [inline lib/moonbeam.js]: no need for auth here, the comeptition will be a public event, no?

File: lib/routes-pub.js (hunk 1 above)
Comment (batrudinych) [reply lib/routes-pub.js]: In order to omit `bind` call wherever you pass the method as a callback. I wanted to see what you think about that.

File: lib/moonbeam.js (hunk 2 above)
Comment (batrudinych) [reply lib/moonbeam.js]: Did constructor binding for public routes only, wanted to check what you think

File: lib/gr-client.js (hunk 3 above)
Comment (batrudinych) [reply lib/gr-client.js]: will do, thanks

File: config/moonbeam.mongo.conf.json.example (hunk 4 above)
Comment (batrudinych) [reply config/moonbeam.mongo.conf.json.example]: yeah, thanks

File: lib/moonbeam.js (hunk 5 above)
Comment (batrudinych) [reply lib/moonbeam.js]: will remove, thank you

File: lib/helpers/grenache.service.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+'use strict'
+
+const path = require('path')
+process.env.EXPRESS_CONFIG_PATH = path.join(process.cwd(), 'config/grenache.service.conf.json')
```
Comment (robertkowalski) [inline lib/helpers/grenache.service.js]: very nice,

maybe we can use `__dirname`, depending on how the service is run and from where process.cwd might be different

File: lib/utils.js [hunk 7]
```diff
@@ -32,11 +32,30 @@ function checkCandlesReq (type, cb) {
...
+  id: val => new RegExp(/^[1-9]\d*$/).test(val),
+  type: val => ['vol', 'pnl'].includes(val)
+}
+exports.verifyCompetitionParams = verifyCompetitionParams
+function verifyCompetitionParams (paramNames = []) {
+  return (req, res, next) => {
+    const suitable = paramNames.every(param => competitionParamsVerifiers[param](req.params[param]))
+    if (suitable) return next()
+
+    res.status(400).json({ error: 'ERR_BAD_REQUEST' })
```
Comment (robertkowalski) [inline lib/utils.js]: nice!

File: lib/helpers/grenache.service.js (hunk 6 above)
Comment (batrudinych) [reply lib/helpers/grenache.service.js]: Oh my god, I have no idea why I put `process.cwd()` in here, of course, it should be `__dirname`, thank you!

**Pull request merged**